lib/connection/pool.js

```diff
@@ -957,64 +994,118 @@ Pool.prototype.write = function(commands, options, cb) {
     cb: cb, raw: false, promoteLongs: true, promoteValues: true, promoteBuffers: false, fullResult: false
   };
 
-  var buffer = null
+  // Ensure commands is an array
+  if (!Array.isArray(commands)) {
+    commands = [commands];
+  }
 
-  if(Array.isArray(commands)) {
-    buffer = [];
+  // For each command, create a promise while we create the command to send to the server
+  var commandsPromises = []
```

Comment:
This sort of dramatically complicates the code imho. The original code is short and sweet:
```
  if(Array.isArray(commands)) {
    buffer = [];

    for(var i = 0; i < commands.length; i++) {
      buffer.push(commands[i].toBin());
    }

    // Get the requestId
    operation.requestId = commands[commands.length - 1].requestId;
  } else {
    operation.requestId = commands.requestId;
    buffer = commands.toBin();
  }
```

Whereas the refactoring here introduces a lot of compression-specific information, as well as forcing everything to use promises - which is generally a no-no in "hot paths", considering the overhead needed to use them. Promises are generally relegated to user-facing API, as their greatest strength is that of readability. 

It looks like the complicated part of this code is building the payload for an `operation`. What I would do is keep as much the same as possible between the sync/async versions of building the operation (things like this: https://github.com/christkv/mongodb-core/blob/2.0/lib/connection/pool.js#L991-L999), isolate the pieces that need to potentially be async and "branch" there - similar to what was done with `decompress`. The pseudocode might look like:

```
  operation.promoteLongs = typeof options.promoteLongs == 'boolean' ? options.promoteLongs : true;
  operation.promoteValues = typeof options.promoteValues == 'boolean' ? options.promoteValues : true;
  operation.promoteBuffers = typeof options.promoteBuffers == 'boolean' ? options.promoteBuffers : false;
  operation.raw = typeof options.raw == 'boolean' ? options.raw : false;
  operation.immediateRelease = typeof options.immediateRelease == 'boolean' ? options.immediateRelease : false;
  operation.documentsReturnedIn = options.documentsReturnedIn;
  operation.command = typeof options.command == 'boolean' ? options.command : false;
  operation.fullResult = typeof options.fullResult == 'boolean' ? options.fullResult : false;
  operation.noResponse = typeof options.noResponse == 'boolean' ? options.noResponse : false;
  ... lots of other operation stuff ....

  if (needToCompress) {
     compress(operation, (err, op) => {
        this.queue.shift(op);
        tryToExecute();
     });
  } else {
    this.queue.shift(op);
    tryToExecute();
  }
```